File: app/src/main/java/com/rockspin/bargainbits/data/models/cheapshark/GameInfo.java
package com.rockspin.bargainbits.data.models.cheapshark;

import java.util.ArrayList;
import java.util.List;

/**
 * response from the game info call.
 */
public class GameInfo {

    private Info info;
    private CheapestPriceEver cheapestPriceEver;
    private List<AbbreviatedDeal> deals = new ArrayList<>();

    public Info getInfo() {
        return info;
    }

    public void setInfo(final Info info) {
        this.info = info;
    }

    public CheapestPriceEver getCheapestPriceEver() {
        return cheapestPriceEver;
    }

    public void setCheapestPriceEver(final CheapestPriceEver cheapestPriceEver) {
        this.cheapestPriceEver = cheapestPriceEver;
    }

    public List<AbbreviatedDeal> getDeals() {
        return deals;
    }

    public void setDeals(final List<AbbreviatedDeal> deals) {
        this.deals = deals;
    }

    public float getLowestSalePrice() {
        float lowestSalePrice = Float.MAX_VALUE;

        for (final AbbreviatedDeal deal : deals) {
            lowestSalePrice = Math.min(deal.getPrice(), lowestSalePrice);
        }

        return lowestSalePrice;
    }

    public float getHighestNormalPrice() {
        float highestNormalPrice = 0;

        for (final AbbreviatedDeal deal : deals) {
            highestNormalPrice = Math.max(deal.getRetailPrice(), highestNormalPrice);
        }

        return highestNormalPrice;
    }

    public float getTopSavingsFraction() {
        float topSavings = 0;

        for (final AbbreviatedDeal deal : deals) {
            topSavings = Math.max(deal.getSavingsFraction(), topSavings);
        }

        return topSavings;
    }

    public boolean isOnSale() {
        for (final AbbreviatedDeal deal : deals) {
            if (deal.getSavingsFraction() > 0.0f) {
                return true;
            }
        }

        return false;
    }

    public boolean containsMultipleStores() {
        return deals.size() > 1;
    }
}

File: app/src/main/kotlin/com/rockspin/bargainbits/data/models/GameInfo.kt
package com.rockspin.bargainbits.data.models

import com.google.gson.annotations.SerializedName

data class GameInfo(
    @SerializedName("info")
    val info: Info,
    @SerializedName("cheapestPriceEver")
    var cheapestPriceEver: CheapestPriceEver? = null,
    @SerializedName("deals")
    val deals: List<AbbreviatedDeal>
) {
    val lowestSalePrice: Double
        get() {
            val lowestSalePrice = deals
                .map { it.price }
                .min()
                ?: Double.MAX_VALUE

            return lowestSalePrice
        }

    val highestNormalPrice: Double
        get() {
            val highestNormalPrice = deals
                .map { it.retailPrice }
                .max()
                ?: 0.0

            return highestNormalPrice
        }

    val topSavingsFraction: Double
        get() {
            val topSavings = deals
                .map { it.savingsFraction }
                .max()
                ?: 0.0

            return topSavings
        }

    val isOnSale: Boolean
        get() {
            return deals.any { it.savingsFraction > 0.0 }
        }
}
